Scatterplot visualization

BinarySoftware opened the pull request:
### Pull Request Description
Creates scatterplot visualization according to task requirements : 
https://app.zenhub.com/workspaces/enso-ide-5e99c2847b8a3027b334f1b9/issues/enso-org/ide/797


<img width="361" alt="Zrzut ekranu 2020-10-22 o 20 04 03" src="https://user-images.githubusercontent.com/26655166/96912395-98751700-14a2-11eb-8958-0c1296d69e72.png">

and dark mode:
<img width="477" alt="Zrzut ekranu 2020-10-23 o 19 46 41" src="https://user-images.githubusercontent.com/26655166/97293301-c95f9e00-184c-11eb-8b92-4f32b5c18560.png">




Example code to generate plot given above:
```ruby
    scatterplotTest = """
                    {"axis" : {"x" : { "label" : "x-axis label", "scale" : "linear" }, "y" : { "label" : "y-axis label", "scale" : "linear" }},
                    "focus" : { "x" : 1.7, "y" : 2.1, "zoom" : 3.0 },
                    "points" : {"labels" : "visible"},
                    "data" : [
                                { "x" : 0.1, "y" : 0.7, "label" : "foo", "color" : "ff0000", "shape" : "circle", "size" : 5.0 },
                                { "x" : 0.1, "y" : 0.7, "label" : "foo", "color" : "ffff00", "shape" : "circle", "size" : 2.0 },
                                { "x" : 0.2, "y" : 0.6, "label" : "bar", "color" : "00ff00", "shape" : "square", "size" : 0.5 },
                                { "x" : 0.3, "y" : 0.4, "label" : "baz", "color" : "0000ff", "shape" : "square", "size" : 0.1 },
                                { "x" : 0.5, "y" : 0.4, "label" : "bao", "color" : "f0000f", "shape" : "triangle", "size" : 1.0 }
                    ]}
```

### Important Notes

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Rust](https://github.com/luna/enso/blob/main/docs/style-guide/rust.md).
- [x] All code has been tested where possible.
- [ ] All code has been profiled where possible.

farmaazon commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -43,111 +71,190 @@ class ScatterPlot extends Visualization {
+                    .attr('transform',d => "translate("+x(d.x)+","+y(d.y)+")")
             }
-        };
+        }
```
This is a really long function. Can you split it into functions instead of making headings?

BinarySoftware commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -43,111 +71,190 @@ class ScatterPlot extends Visualization {
-
-            var color = d3.scaleOrdinal()
-                .domain(colorDomain)
-                .range(colorRange)
-
-            var brush = d3.brushX()
-                .extent([[0, 0], [width, height]])
-                .on("end", updateChart)
-
-            var scatter = svg.append('g')
-                .attr("clip-path", "url(#clip)")
+        /////////////
 
-            scatter
-                .selectAll("circle")
-                .data(_data)
+        let y = d3.scaleLinear()
+        if(axis.y.scale !== "linear") {
+            y = d3.scaleLog();
+        }
+
+        y.domain([yMin - dy, yMax + dy])
+            .range([height, 0]);
+        svg.append("g")
+            .call(d3.axisLeft(y));
```
because we zoom only on x axis, tried to do zooming by brushing on both axes but didn't work

farmaazon commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -32,122 +45,200 @@ class ScatterPlot extends Visualization {
+
+        this.createLabels(axis, svg, box_width, margin, box_height);
+
+        let scatter = this.createSymbols(svg, box_width, box_height, points, dataPoints, x, y);
```
So, scatter is a structure with symbols? If yes, I would make `scatter = this.createScatter` or `symbols = this.createSymbols` (I prefer the former).

farmaazon commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -32,122 +45,200 @@ class ScatterPlot extends Visualization {
+            if (points.labels === "visible") {
                 scatter
-                    .selectAll("circle")
+                    .selectAll("text")
                     .transition().duration(1000)
-                    .attr("cx", function (d) {
-                        return x(d[headerNames[0]]);
-                    })
-                    .attr("cy", function (d) {
-                        return y(d[headerNames[1]]);
-                    })
-
+                    .attr('transform', d => "translate(" + x(d.x) + "," + y(d.y) + ")")
```
The `transition().duration(...).attr(...) pattern seems repetitive, maybe make some lambda from it?

BinarySoftware replied:
removed as other zooming method will be used.

farmaazon reviewed (commented):
![image](https://user-images.githubusercontent.com/3919101/94449972-f9177980-01ac-11eb-9ce2-90f676a4d0eb.png)

On the example you provided the labels on x are often overlapping each other.

farmaazon reviewed (changes requested):
> - User should be able to interactively zoom and pan visualization, using the same mouse navigation patterns as in graph editor.
> 
> - Left mouse button should be used for selection: https://www.d3-graph-gallery.com/graph/interactivity_brush.html . The selection will not do anything important right now, but this visualization should be also used as widget in the future, which means, that after selecting few points user could get that data passed to the next node.
> 
> - In the top right corner, there should be an icon which clicked should zoom the graph this way that all points are visible.
> 
> - After points are selected, there should be a shortcut (let it be "s") which zooms only to the selected region. The shortcut "a", should show all points again. After points are selected, there should also appear an icon on the right top corner of the visualization (below the "zoom to show all" icon), which clicked, will zoom the selected points.

Above points seem to be not addressed:
* zooming and panning works, but not as in graph editor
* There is no selection nor icon to unzoom, nor shortcut for zooming selected points.

Additionally, please implement initial focus as discussed on DM.

BinarySoftware commented:
> 
> Additionally, please implement initial focus as discussed on DM.

Done

farmaazon reviewed (changes requested):
1. The fit window went outside the visualization
2. the labels animate strangely after "fitting all" - but that's actually is not a big deal
3. sometimes when panning right after "fitting all" the plot "jumps".
I sent it in DM, because attaching recording here didn't work for some reason.

BinarySoftware commented:
> * The fit window went outside the visualization
> * the labels animate strangely after "fitting all" - but that's actually is not a big deal
> * sometimes when panning right after "fitting all" the plot "jumps".
>   I sent it in DM, because attaching recording here didn't work for some reason.

Fixed all

farmaazon reviewed (changes requested):
* missing shortcut for fitting to selected points
* The brush area is scrolled with points

wdanilo commented:
Let's start with fixing visual elements first:
<img width="215" alt="img1" src="https://user-images.githubusercontent.com/1623053/95215417-3d2ffd00-07f1-11eb-9010-ffbf96b6317c.png">

1. Red rectangle - the label is too close to the numbers. The spacing should be the same as spacing on the second axis. Moreover, the spacing should be configurable with a variable inside of the js code.
2. Green rectangle - the labels are overlapping. Please be sure it doesnt happen.
3. Dark green rectangle - the labels look very bad on top of the points. Make them on the right side instead, like here: https://observablehq.com/@d3/scatterplot
4. Blue rectangle - the label is not axis-centered
5. pink rectangle - acording to the spec "In the top right corner, there should be an icon which clicked should zoom the graph this way that all points are visible."
6. The font is different than the font we use in GUI, thus it looks bad.

Moreover, I would like to see on a gif (several pics) if everything from the sec is implemented, in particular: 
- "border-color", "border-size", different sizes are working correctly.
- "User should be able to interactively zoom and pan visualization, using the same mouse navigation patterns as in graph editor."
- "After points are selected, there should be a shortcut (let it be "s") which zooms only to the selected region. The shortcut "a", should show all points again. After points are selected, there should also appear an icon on the right top corner of the visualization (below the "zoom to show all" icon), which clicked, will zoom the selected points."

etc. Please be sure that EVERYTHING is implemented from the spec and show it on gif / pics before code review.

wdanilo reviewed (changes requested):
Described above

BinarySoftware commented:
> Let's start with fixing visual elements first:
> <img alt="img1" width="215" src="https://user-images.githubusercontent.com/1623053/95215417-3d2ffd00-07f1-11eb-9010-ffbf96b6317c.png">
> 
> 1. Red rectangle - the label is too close to the numbers. The spacing should be the same as spacing on the second axis. Moreover, the spacing should be configurable with a variable inside of the js code.
> 2. Green rectangle - the labels are overlapping. Please be sure it doesnt happen.
> 3. Dark green rectangle - the labels look very bad on top of the points. Make them on the right side instead, like here: https://observablehq.com/@d3/scatterplot
> 4. Blue rectangle - the label is not axis-centered
> 5. pink rectangle - acording to the spec "In the top right corner, there should be an icon which clicked should zoom the graph this way that all points are visible."
> 6. The font is different than the font we use in GUI, thus it looks bad.
> 
> Moreover, I would like to see on a gif (several pics) if everything from the sec is implemented, in particular:
> 
> * "border-color", "border-size", different sizes are working correctly.
> * "User should be able to interactively zoom and pan visualization, using the same mouse navigation patterns as in graph editor."
> * "After points are selected, there should be a shortcut (let it be "s") which zooms only to the selected region. The shortcut "a", should show all points again. After points are selected, there should also appear an icon on the right top corner of the visualization (below the "zoom to show all" icon), which clicked, will zoom the selected points."
> 
> etc. Please be sure that EVERYTHING is implemented from the spec and show it on gif / pics before code review.
> 
> > <img alt="" width="48" src="https://camo.githubusercontent.com/e381d73cc7f82cc50a41f278c2e68cb6a6257ef2/68747470733a2f2f7374617469632e6f627365727661626c6575736572636f6e74656e742e636f6d2f7468756d626e61696c2f333634353965636631373339643466653666643461323131333130393166643664356361333432313837393162353530333732613362386633666463326162352e6a7067">**[Scatterplot](https://observablehq.com/@d3/scatterplot)**An Observable notebook by D3.

1. - fixed
2. - fixed
5. - (as said in DM) - this one needs to be here, otherwise it is shadowed by vis-selector, and cannot be clicked.
6. - fixed

Size and color : 
<img width="227" alt="Zrzut ekranu 2020-10-12 o 16 05 31" src="https://user-images.githubusercontent.com/26655166/95755595-e751ce00-0ca4-11eb-9e44-90840e99255a.png">

TODO: 
- Fix label position near shape
- axis-center legend
- shortcuts

farmaazon commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -108,41 +123,81 @@ class ScatterPlot extends Visualization {
 
-            scaleAndAxis.xAxis.call(d3.axisBottom(new_xScale).ticks(7));
-            scaleAndAxis.yAxis.call(d3.axisLeft(new_yScale).ticks(7));
+            scaleAndAxis.xAxis.call(d3.axisBottom(new_xScale).ticks(box_width/30));
```
Magic number.

farmaazon commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -108,41 +123,81 @@ class ScatterPlot extends Visualization {
             if (points.labels === "visible") {
                 scatter.selectAll("text")
-                    .attr('transform', d => "translate(" + new_xScale(d.x) + "," + new_yScale(d.y) + ")")
+                    .attr("x", d => new_xScale(d.x) + 7)
+                    .attr("y", d => new_yScale(d.y) + 2)
```
Magic numbers.

farmaazon commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -108,41 +123,81 @@ class ScatterPlot extends Visualization {
-            let extent = d3.event.selection
-            scatter.classed("selected", d => isBrushed(extent, scaleAndAxis.xScale(d.x), scaleAndAxis.yScale(d.y) ))
+        function zoomin() {
+            scaleAndAxis.xScale.domain([scaleAndAxis.xScale.invert(extent[0][0]), scaleAndAxis.xScale.invert(extent[1][0])]);
+            scaleAndAxis.yScale.domain([scaleAndAxis.yScale.invert(extent[1][1]), scaleAndAxis.yScale.invert(extent[0][1])]);
```
Long lines.

farmaazon commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -108,41 +123,81 @@ class ScatterPlot extends Visualization {
+
+            scatter.selectAll("path")
+                .transition().duration(1000)
+                .attr('transform', d => "translate(" + scaleAndAxis.xScale(d.x) + "," + scaleAndAxis.yScale(d.y) + ")")
```
Long line.

farmaazon commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -108,41 +123,81 @@ class ScatterPlot extends Visualization {
+            scaleAndAxis.yScale.domain([scaleAndAxis.yScale.invert(extent[1][1]), scaleAndAxis.yScale.invert(extent[0][1])]);
+
+            scaleAndAxis.xAxis.transition().duration(1000)
+                .call(d3.axisBottom(scaleAndAxis.xScale).ticks(box_width/30));
```
Magic number.

farmaazon commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -108,41 +123,81 @@ class ScatterPlot extends Visualization {
+            if (points.labels === "visible") {
+                scatter.selectAll("text")
+                    .transition().duration(1000)
+                    .attr("x", d => scaleAndAxis.xScale(d.x) + 7)
+                    .attr("y", d => scaleAndAxis.yScale(d.y) + 2)
```
Magic numbers.

farmaazon commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -172,19 +227,19 @@ class ScatterPlot extends Visualization {
                 else if (d.shape === "triangle") { return d3.symbolTriangle }
                 else                             { return d3.symbolCircle   }
-            }))
+            }).size(d => (d.size || 1.0) * 100))
```
magic number.

farmaazon commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -172,19 +227,19 @@ class ScatterPlot extends Visualization {
                 .append("text")
                 .text(d => d.label)
-                .attr('transform', d => "translate(" + scaleAndAxis.xScale(d.x) + "," + scaleAndAxis.yScale(d.y) + ")")
+                .attr("x", d => scaleAndAxis.xScale(d.x) + 7)
+                .attr("y", d => scaleAndAxis.yScale(d.y) + 2)
```
Magic numbers.

farmaazon commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -225,13 +288,15 @@ class ScatterPlot extends Visualization {
             .attr("transform", "translate(0," + box_height + ")")
-            .call(d3.axisBottom(xScale).ticks(7))
+            .attr("style", label_style)
+            .call(d3.axisBottom(xScale).ticks(box_width/30))
```
Magic number.

farmaazon commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -342,7 +413,7 @@ class ScatterPlot extends Visualization {
 
             scaleAndAxis.xAxis.transition().duration(1000)
-                .call(d3.axisBottom(scaleAndAxis.xScale).ticks(7));
+                .call(d3.axisBottom(scaleAndAxis.xScale).ticks(box_width/30));
```
Magic number.

farmaazon commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -353,14 +424,31 @@ class ScatterPlot extends Visualization {
                 scatter.selectAll("text")
                     .transition().duration(1000)
-                    .attr('transform', d => "translate(" + scaleAndAxis.xScale(d.x) + "," + scaleAndAxis.yScale(d.y) + ")")
+                    .attr("x", d => scaleAndAxis.xScale(d.x) + 7)
+                    .attr("y", d => scaleAndAxis.yScale(d.y) + 2)
```
Magic numbers.

farmaazon reviewed (changes requested):
There is some issue with visiblity of "Zoom to selected" button: it is shown when I click "Fit to all" button, and is not hidden when deselecting points. Actually the only way to hide it is pressing ctrl+s.

Additionally, there are many places in code annotated with "Magic number(s)" - please move them to named constants.

farmaazon reviewed (changes requested):
when I click "Zoom to selected" I get error in console:
```Uncaught TypeError: self.zoomingHelper is not a function
    at HTMLButtonElement.zoomIn (eval at imports.wbg.__wbg_newwithargs_3e8c495287591add (eval at <anonymous> (index.js:155)), <anonymous>:167:18)```
After clicking "Fit all" I can start brushing only on half of the visible plot.

farmaazon commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -97,9 +102,10 @@ class ScatterPlot extends Visualization {
+        this.addBrushing(box_width, box_height, scatter, scaleAndAxis, selectedZoomBtn, points, zoom);
     }
 
+    // TODO: Passthrough events for right-click and scroll from brush to zoom.
```
```suggestion
```

farmaazon commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -142,16 +144,20 @@ class ScatterPlot extends Visualization {
         let extent;
-        let brush = d3.brush()
-            .extent([[0, 0], [box_width, box_height]])
+        let brush = d3.brush().filter(function() {
+            return d3.event.type === "mousedown" && d3.event.button === 0
+        }).extent([[0, 0], [box_width, box_height]])
```
```suggestion
        let brush = d3.brush()
            .extent([[0, 0], [box_width, box_height]])
```

wdanilo reviewed (changes requested):
The styles are wrong, as we described in DM

wdanilo commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -1,154 +1,372 @@
+            brushElem.call(brush.move, null);
+            selectedZoomBtn.style.display = "none";
+            selectedZoomBtn.removeEventListener("click",zoomIn,true)
+            document.removeEventListener('keydown', zoomInKeyEvent,true);
```
all the styles in this file are inconsistent - sometimes there are spaces around `,`, sometimes not. Just apply Rust style guide here :)

wdanilo commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -1,154 +1,372 @@
-            var brush = d3.brushX()
-                .extent([[0, 0], [width, height]])
-                .on("end", updateChart)
+        document.addEventListener('click'      , endBrushing,false);
+        document.addEventListener('auxclick'   , endBrushing,false);
+        document.addEventListener('contextmenu', endBrushing,false);
+        document.addEventListener('scroll'     , endBrushing,false);
```
refactor that to a loop

wdanilo commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -1,154 +1,372 @@
-            var scatter = svg.append('g')
-                .attr("clip-path", "url(#clip)")
+    zoomingHelper(scaleAndAxis, box_width, scatter, points) {
+        scaleAndAxis.xAxis.transition().duration(1000)
```
everywhere hardcoded literals. Il;l not post more about it - just refactor all hardcoded literals to appropriate section with constants on top of the file.

wdanilo commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -1,154 +1,372 @@
+        }
+    }
+
+    createScatter(svg, box_width, box_height, points, dataPoints, scaleAndAxis) {
```
this function is really really long. It needs either sub-sections inside of the body, or divide it to smaller helper init functions

wdanilo commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -159,10 +377,97 @@ class ScatterPlot extends Visualization {
+            padding: 2px 10px;
+            outline: none;
+            background-color: transparent;
+            border: 1px solid #333;
+            color: #333;
```
all collors that are used in multiple places should als o be refactored to vars

BinarySoftware commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -1,154 +1,372 @@
+
+        let selectedZoomBtn = this.createButtonScaleToPoints();
+
+        this.addBrushing(box_width, box_height, scatter, scaleAndAxis, selectedZoomBtn, points, zoom);
```
Yes they are <3 why is this bad? It is more readable that way imo

BinarySoftware commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -1,154 +1,372 @@
+
+        let self = this;
+        function zoomIn() {
+            let xMin = scaleAndAxis.xScale.invert(extent[0][0]);
+            let xMax = scaleAndAxis.xScale.invert(extent[1][0]);
+            let yMin = scaleAndAxis.yScale.invert(extent[1][1]);
+            let yMax = scaleAndAxis.yScale.invert(extent[0][1]);
```
i'll address that on tommorrow's call :)

wdanilo replied:
This is still not fixed in the code. The reader of the code has to understand that. If this is the proper solution, could you add explanation WHY plus maybe add some links to the docs?

BinarySoftware commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -1,154 +1,372 @@
+            .attr("class", "brush")
+            .call(brush)
+
+        let self = this;
+        function zoomIn() {
```
This is a helper function used in 2 places, and exists like that to be able to remove event listeners when nothing is selected.

wdanilo commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -1,171 +1,560 @@
+/**
+ * Helper function to load styles.
+ *
+ * It runs only once, so it won't load the same style on this or any other visualization show/hide.
+ */
+function loadStyle(url) {
+    let style   = document.createElement("link");
+    style.href  = url;
+    style.rel   = "stylesheet";
+    style.media = "screen";
+    style.type  = "text/css";
+
+    document.head.appendChild(style);
+}
```
Where in the code it is guaranteed it runs only once? I see we call it on every incoming data, where is the check?

farmaazon replied:
No. The code is run once per loading of visualization. On every incoming data the onDataReceived is called.

BinarySoftware replied:
Ditto: Actually the "before" implementation worked perfectly because the file was loaded only once, and then, the only thing that got called was the onDataReceived. So by implementation it is and must be load only once.

wdanilo replied:
Ok, that makes sense, but now I dont understand the docs here. The docs here tell that it runs only once, but this is not guaranteed in the body. So the docs are wrong now.

wdanilo commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -1,171 +1,560 @@
+
+const label_style        = "font-family: DejaVuSansMonoBook; font-size: 10px;";
+const x_axis_label_width = 30;
+const label_padding_x    = 7;
+const label_padding_y    = 2;
```
above you have var `x_asis_label_...` which tells what kind of label this is. Are these vars applied to every label, bot x and y?

wdanilo commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -1,171 +1,560 @@
+     * Presents a scatterplot visualization after receiving `data`.
+     */
     onDataReceived(data) {
+        loadScript('https://d3js.org/d3.v4.min.js');
+        loadStyle('https://fontlibrary.org/face/dejavu-sans-mono')
```
Are ou loading the script / style on every new data received? This is even worse than it was before. If not, where is the check that it does get loaded once? I havent seen that in the code.

BinarySoftware replied:
Actually the "before" implementation worked perfectly because the file was loaded only once, and then, the only thing that got called was the onDataReceived. So by implementation it is and must be load only once.

wdanilo commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -1,171 +1,560 @@
-        const divElem = this.createDivElem(width, height);
+        let width        = this.dom.getAttributeNS(null,"width");
+        let height       = this.dom.getAttributeNS(null,"height");
+        const btnPadding = 25;
```
what kind of apdding this is? The var does not tell wherther this is top, bottom padding and from what element

wdanilo commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -1,171 +1,560 @@
+        this.dom.appendChild(divElem);
+
+        let parsedData = JSON.parse(data);
+        let axis       = parsedData.axis || {x:{scale:"linear"},y:{scale:"linear"}};
```
This PR still contains a lot of literals that are repeating. For example, the string "linear" repeats in this code 4 times. Please refactor all common literals to constants. Of course, this does not apply to CSS variables.

farmaazon replied:
I disagree with this very example. But is there a point for making constants like `const linear_scale = "linear"`? It does not tell much, and will be checked for errors in runtime anyway.

BinarySoftware replied:
I kinda agreed with two examples that were repeated more than 3 times and refactored them.

wdanilo commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -1,171 +1,560 @@
-            });
+            .attr("transform","translate(" + margin.left + "," + margin.top + ")");
+
+        let extremesAndDeltas = this.getExtremesAndDeltas(dataPoints);
+
+        let scaleAndAxis = this.createAxes(axis,extremesAndDeltas,box_width,box_height,svg,focus);
+
+        this.createLabels(axis,svg,box_width,margin,box_height);
+
+        let scatter = this.createScatter(svg,box_width,box_height,points,dataPoints,scaleAndAxis);
+
+        let zoom = this.addPanAndZoom(box_width,box_height,svg,margin,scaleAndAxis,scatter,points);
+
+        // TODO: Visualization selector obfuscates button, so it is now on the bottom, should be on top.
+        this.createButtonFitAll(scaleAndAxis,scatter,points,extremesAndDeltas,zoom,box_width);
+
+        let selectedZoomBtn = this.createButtonScaleToPoints();
+
+        this.addBrushing(box_width,box_height,scatter,scaleAndAxis,selectedZoomBtn,points,zoom);
```
I had here comments in my previous pr about newlines - they are not applied nor discussion is done here

wdanilo commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -1,171 +1,560 @@
+
+        let zoom = this.addPanAndZoom(box_width,box_height,svg,margin,scaleAndAxis,scatter,points);
+
+        // TODO: Visualization selector obfuscates button, so it is now on the bottom, should be on top.
```
make the todo proper one, according to our rules (who is the author, what is happening right now, how it sohuld be changed, why it cant bechanged right now, link to issue)

wdanilo commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -1,171 +1,560 @@
+     * Adds panning and zooming functionality to the visualization.
+     */
+    addPanAndZoom(box_width,box_height,svg,margin,scaleAndAxis,scatter,points) {
+        const extent = [.5,20];
```
magical numbers - what are these? should be extracted to well named variables.

BinarySoftware replied:
Refactored them, but just to overspecify, this is how zooming is defined in almost every d3js pan/zoom example out there.

wdanilo commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -1,171 +1,560 @@
+        }
+
+        const zoomInKeyEvent = (event) => {
+            if ((event.ctrlKey || event.metaKey) && event.key === 'z') {
```
refactor all these shortcuts to a shortcut section on top of file (below sconstants section). Like:

```js
let shortcuts = {
    zoomIn : (e) => ((e.ctrlKey || e.metaKey) && e.key === 'z'),
    ...
}
```

wdanilo commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -1,171 +1,560 @@
+        return {xScale:xScale,yScale:yScale,xAxis:xAxis,yAxis:yAxis};
+    }
+
+    /**
+     * Helper function calculating domains of given data.
+     */
```
this docs does not tell anything. What is "domain" what is "given data"? Need to be better descirbed, preferably with examples

wdanilo commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -1,171 +1,560 @@
+    }
+
+    /**
+     * Helper function to get text width to make sure text wont overlap on screen.
```
overlap with what?

wdanilo commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -1,171 +1,560 @@
-        divElem.setAttributeNS(null, "height", "100%");
-        divElem.setAttributeNS(null, "transform", "matrix(1 0 0 -1 0 0)");
+    /**
+     * Helper function for zooming into given scale.
```
this doc doesnt explain what this function really does basedo n the input parameters. Please fix this and other docs - every doc should be very clear of what it does and why it needs the given inputs. Like here - this is a zooming into "given" scale. But the function does not consume a "scale", so how its given? Instead it conumes tons of strange parameters whose purpose is completely not mentioned in the docs. Please fix the docs in such way, that every function is clear about how it uses input parameters to compute something.

wdanilo commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -1,171 +1,560 @@
+        let scatter = svg.append('g')
+            .attr("clip-path","url(#clip)")
+
+        let size_scale = 100
```
what size scale this is about? What does this variable mean? What would happen if I set it to 1? It should be clear what it is and what it is used for.

BinarySoftware replied:
This is a multiplier so that points are visible on plot 😂 as the value specified in the issue is really small, compared to what is necessary

wdanilo commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -1,171 +1,560 @@
+     */
+    matchShape() {
+        return d => {
+            if (d.shape === undefined)  { return d3.symbolCircle   }
+            if (d.shape === "cross")    { return d3.symbolCross    }
+            if (d.shape === "diamond")  { return d3.symbolDiamond  }
+            if (d.shape === "square")   { return d3.symbolSquare   }
+            if (d.shape === "star")     { return d3.symbolStar     }
+            if (d.shape === "triangle") { return d3.symbolTriangle }
+            return d3.symbolCircle
```
The first if is not needed.

wdanilo commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -1,171 +1,560 @@
+    }
+
+    /**
+     * Helper function calculating extreme values and paddings to make sure data will fit nicely.
```
Again (as with other docs) - what does it mean? Cool the data will fit nicely, but you need to describe what this function does with the data and HOW it computes the things. When someone else is reading the code, they need to understan the logic behind a function to know if it suits their needs.

wdanilo commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -1,171 +1,594 @@
+        function wheeled() {
+            let current_transform = d3.zoomTransform(scatter);
+            if (d3.event.ctrlKey) {
+                current_transform.k = current_transform.k - d3.event.deltaY * 0.01;
```
what is this magical `0.01`? Should be constant

wdanilo commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -1,171 +1,594 @@
+         */
+        const endBrushing = function (_) {
+            brushElem.call(brush.move,null);
+            selectedZoomBtn.style.display = "none";
+            selectedZoomBtn.removeEventListener("click",zoomIn,true)
```
these lines are duplicated in this codebase, while they just mean "hideZoomButton". They should be refactored to separate function. The same applies to hiding / showing other buttons if there is other duplicated code

wdanilo commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -1,171 +1,594 @@
+            .attr("clip-path","url(#clip)")
+
+        let sizeScaleMultiplier = 100
+
```
You put enters inside of the body, which means the body should either be refactored to separate functions or sections should be added here. I understand why you added newlines - these things are really long here. And a section / refactoring would help a lot here. When reading the code its hard for me to understand what then ext 10 lines do - just add section with explanation! Every time you need to put enter inside of a body, you need to use sections or refactoring.

wdanilo commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -1,171 +1,594 @@
+     * Helper function calculating domains for x and y axes.
+     *
+     * Example:
+     * Lets say we have a bunch of points. Those points will have some minimum and maximum value,
+     * from which we can calculate the span of points on X and Y axis, hence domain, with added
+     * padding to make sure points will fit nicely on the chart.
+     */
```
That are amazing docs, good job <3

wdanilo commented on src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js:
```diff
@@ -1,171 +1,594 @@
+
+        let self = this;
+        function unzoom() {
+            zoom.zoomElem.transition().duration(0).call(zoom.zoom.transform,d3.zoomIdentity);
```
is this diabling unzoom animation? If so, let it be there! We love animations, we hate "view jumps"! :)
